packages/cosmic-swingset/src/launch-chain.js

```diff
@@ -259,25 +280,57 @@ export async function buildSwingset(
  */
 
 /**
- * @param {BeansPerUnit} beansPerUnit
+ * @param {object} params
+ * @param {BeansPerUnit} params.beansPerUnit
+ * @param {import('@agoric/swingset-vat').CleanupBudget} [params.vatCleanupBudget]
  * @param {boolean} [ignoreBlockLimit]
  * @returns {ChainRunPolicy}
  */
 function computronCounter(
-  {
+  { beansPerUnit, vatCleanupBudget },
+  ignoreBlockLimit = false,
+) {
+  const {
     [BeansPerBlockComputeLimit]: blockComputeLimit,
     [BeansPerVatCreation]: vatCreation,
     [BeansPerXsnapComputron]: xsnapComputron,
-  },
-  ignoreBlockLimit = false,
-) {
+  } = beansPerUnit;
   assert.typeof(blockComputeLimit, 'bigint');
   assert.typeof(vatCreation, 'bigint');
   assert.typeof(xsnapComputron, 'bigint');
+
   let totalBeans = 0n;
-  const shouldRun = () => ignoreBlockLimit || totalBeans < blockComputeLimit;
-  const remainingBeans = () =>
-    ignoreBlockLimit ? undefined : blockComputeLimit - totalBeans;
+  const { default: defaultCleanupBudget = 0, ...remainingCleanups } = {
+    ...vatCleanupBudget,
+  };
+  let cleanupStarted = false;
+  // Cleanup starts off "done" unless at least one budget field is positive.
+  let cleanupDone = !Object.values(remainingCleanups).some(n => n > 0);
+  // We currently implement a strict separation: first a phase in which
+  // deliveries are allowed but cleanup is not, then (iff no work was done in
+  // the first) a phase in which cleanup is allowed but deliveries are not.
+  // https://github.com/Agoric/agoric-sdk/issues/8928#issuecomment-2053357870
+  const shouldRun = () =>
+    !cleanupStarted && (ignoreBlockLimit || totalBeans < blockComputeLimit);
+  const allowCleanup = () => cleanupStarted && !cleanupDone;
```

Comment:
This won't get us the right budget.. `allowCleanup() === true` means an unlimited budget, and the kernel only calls `didCleanup` at the end of the crank. So a non-finite budget will let the kernel perform unlimited cleanup, and then it will report a huge `didCleanup(cleanups)`, and our decrementer will stop it, but by then it will be too late.

We need `allowCleanup` to return the remaining budget each time (if we're in the cleanup window). We also need to let the kernel perform GC actions and BOYD. The kernel will ask about one vat with `allowCleanup()`, do an amount of cleanup work limited by the returned budget, notify the policy with a `didCleanup()`, then it returns to the "decide what work to do next" function, which will query the policy again. We need the `didCleanup()` to return `true` (meaning "keep going, don't stop `controller.run()` yet"), then that second call to `allowCleanup()` must say "no more cleanups", to allow it to move on to the gc-action queue, or the dirty-vat-do-BOYD check. We still count computrons, so if e.g. the cleanup resulted in multiple vats becoming dirty and eligible for BOYD, we might perform one BOYD, decide the computrons it spent were too much, and then end the block before we get to the second BOYD. The next block's `crankerPhase = 'leftover'` run will finish off those BOYDs.

This only really works because we only do cleanup on empty blocks, so we know the run-queue is empty before clean is allowed to proceed. The run-policy language doesn't have a way to say "do the gcActions and BOYDs triggered by cleanup, but don't allow any *normal* cranks to be delivered, because we're only doing cleanup-related stuff now".

(I wish there were a good way to unit test this)

So something more like:

```js
let inCleanup = false;
const startCleanup = () => if (totalBeans > 0n) { return false; } else { inCleanup = true; return true; };
const allowCleanup = () => inCleanup && remainingCleanups; // returns false or budget
const didCleanup = details => { .. decrement remainingCleanups; return true; };
const shouldRun = () => ignoreBlockLimit || totalBeans < blockComputeLimit;
```

Reply:
Oof, yeah. Updated to something that actually has the right return type for `allowCleanup`.